api/next.pb.txt

```diff
@@ -5359,6 +5359,20 @@ file {
       type_name: ".google.protobuf.Any"
       json_name: "options"
     }
+    field {
+      name: "export_to_archive"
```

Comment:
These changes don't feel right.
I feel like this should be handled inside the cri plugin rather than pushing it to the task service.

What's the reasoning behind adding to the API?

Reply:
Not sure I can answer it. I just tried to make it work.

On the most inner level of the call stack, where the actual checkpoint is created, I need to know what kind of checkpoint archive it is. The current setup does not enable me to pass this information from the the place the CRI call is decoded to the place where the checkpoint archive is created.

> What's the reasoning behind adding to the API?

That's how I was able to make it work. Sorry. I cannot provide a better answer.

Reply:
Ah, okay.

For Kubernetes checkpointing I am writing the checkpoint to an archive file from which the container can be later restored. To create the checkpoint archive as passed in the CRI I need the additional parameters.

What I actually would like to have is a standard checkpoint archive as suggested in https://github.com/opencontainers/image-spec/issues/962, but as long that is not working I am using my own defined format as used in Podman and CRI-O. The problem with the containerd specific format that already exists is that it includes containerd specific protobuf blobs which make it not usable outside of containerd. I was actually looking at the containerd specific OCI images, but was not able to use them outside of containerd.

My goal is to have the checkpoint files as created by CRIU and some additional JSON files, as I currently do in Podman and CRI-O, but as a standard everybody can use and not just something I came up with. That way any container engine should be able to restore a container checkpointed with another engine. I had success with restoring Podman checkpoints in CRI-O, so it is doable.

The two options I added here `export_to_archive` and `location` are used to tell that an external checkpoint archive should be created and in which file (`location`) it should be created. This is what Kubernetes expects.

If you do not want my self defined container archive format in containerd this PR can be closed as long as the standardizing discussion is unfinished. I can totally understand that that might be undesirable. I just wanted to offer the same support of Kubernetes CRI checkpointing as I have added to CRI-O.

The parameters to the CRI RPC CheckpointContainerRequest() are limited to the container id, location and timeout.

If you feel this is not ready I am okay closing it.

Reply:
/cc @containerd/maintainers 

I'm a bit conflicted here.
On the one hand if other runtimes are supporting a given format I think it's fine to pick it up and perhaps that will even influence the discussions in OCI.
On the other hand adding another bespoke format is not ideal.

If we do add the format, I think it would be best to pass the desired format down but not the other options (export_to_archive and location). CRI can handle the k8s side of it, like maybe instead of passing it the normal content store cri could have it's own store interface for writing it to disk as needed by k8s.